### PR TITLE
Remove the old AST-based backend from rustc_trans.

### PR DESCRIPTION
**Pull request by eddyb**
Starting with Rust 1.13, `--disable-orbit` , `-Z orbit=off` and `#[rustc_no_mir]` have been removed.
Only the new MIR backend is left in the compiler, and only early const_eval uses ASTs from other crates.

Filling drop (previously "zeroing drop"), `#[unsafe_no_drop_flag]` and associated unstable APIs are gone.
Implementing `Drop` doesn't add a flag anymore to the type, all of the dynamic drop is function local.
This is a [breaking-change], please use `Option::None` and/or `mem::forget` if you are unsure about your ability to prevent/control the drop of a value. In the future, `union` will be usable in some such cases.

**NOTE**: DO NOT MERGE before we get the new beta as the stage0, there's some cruft to remove.

All of this will massively simplify any efforts to implement (and as such it blocks) features such as `union`s, safe use of `#[packed]` or new type layout optimizations, not to mention many other experiments.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Dat diff.

**Comment by nrc** [comment]
Much as I'd love to r+ this, r? @nikomatsakis

File: src/librustc_trans/adt.rs [hunk 1]
```diff
@@ -194,10 +162,6 @@ pub fn represent_type<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
     repr
 }
 
-fn dtor_to_init_u8(dtor: bool) -> u8 {
-    if dtor { 1 } else { 0 }
-}
-
 pub trait GetDtorType<'tcx> { fn dtor_type(self) -> Ty<'tcx>; }
```
Comment (jld) [inline src/librustc_trans/adt.rs]: It looks like this trait can also be deleted?

**Comment by eddyb** [comment]
FWIW, this appears to fix #35408 locally, possibly due to `P::map` not using memset with `0x1d` to prevent drop on panic, allowing LLVM to optimize the moves in the AST folder better.

File: src/liballoc/arc.rs [hunk 2]
```diff
@@ -755,11 +756,12 @@ impl<T: ?Sized> Drop for Weak<T> {
     /// ```
     fn drop(&mut self) {
         let ptr = *self.ptr;
-        let thin = ptr as *const ();
 
         // see comments above for why this check is here
-        if thin as usize == mem::POST_DROP_USIZE {
-            return;
+        #[cfg(stage0)] {
```
Comment (nikomatsakis) [inline src/liballoc/arc.rs]: this is nifty

File: src/libcore/intrinsics.rs [hunk 3]
```diff
@@ -244,19 +244,6 @@ extern "rust-intrinsic" {
...
-    ///
-    /// `init_dropped` is unsafe because it returns a datum with all
-    /// of its bytes set to the drop flag, which generally does not
-    /// correspond to a valid value.
-    ///
-    /// This intrinsic is likely to be deprecated in the future when
-    /// Rust moves to non-zeroing dynamic drop (and thus removes the
-    /// embedded drop flags that are being established by this
-    /// intrinsic).
-    pub fn init_dropped<T>() -> T;
```
Comment (nikomatsakis) [inline src/libcore/intrinsics.rs]: buh-bye 👋

File: src/libcore/mem.rs [hunk 4]
```diff
@@ -542,31 +517,20 @@ macro_rules! repeat_u8_as_u64 {
...
-pub const POST_DROP_U64: u64 = repeat_u8_as_u64!(POST_DROP_U8);
+#[cfg(stage0)]
+const POST_DROP_U8: u8 = 0x1d;
 
-#[cfg(target_pointer_width = "16")]
-#[unstable(feature = "filling_drop", issue = "5016")]
-#[allow(missing_docs)]
-pub const POST_DROP_USIZE: usize = POST_DROP_U16 as usize;
-#[cfg(target_pointer_width = "32")]
+// TODO HACK FIXME(eddyb) !!! Remove POST_DROP_* after beta branches off.
```
Comment (nikomatsakis) [inline src/libcore/mem.rs]: I'm confused --- what makes this a hack?

File: src/librustc/ty/layout.rs [hunk 5]
```diff
@@ -891,17 +891,6 @@ impl<'a, 'gcx, 'tcx> Layout {
...
-                // won't actually be in the location we say it is because it'll be after
-                // the unsized field. Several other pieces of code assume that the unsized
-                // field is definitely the last one.
-                if def.dtor_kind().has_drop_flag() &&
-                   ty.is_sized(tcx, &infcx.parameter_environment, DUMMY_SP) {
-                    st.extend(dl, Some(Ok(&Scalar {
-                        value: Int(I8),
-                        non_zero: false
-                    })).into_iter(), ty)?;
-                }
```
Comment (nikomatsakis) [inline src/librustc/ty/layout.rs]: 🐊 tears

File: src/libcore/mem.rs (hunk 4 above)
Comment (eddyb) [inline src/libcore/mem.rs]: That's just an anti-tidy marker so we remove that stuff before merging. Which is why we need to update the stage0 first.

**Pull request merged**